### ISSUE TITLE
Websocket server does not detect broken link on linux

### ISSUE THREAD

**Issue by warenbe** [deniszykov/WebSocketListener#31]
Hi 
i mainly use the server on linux (centos 7 & debian 8) with mono (latest version)
i have a strange problem only when server is running on linux, and not on windows

a user connect using Chrome or firefox to the server. 
and after few minutes, the link is broken (3g lost, wifi lost, etc)
on windows, after few secondes, i  get an exception on the writeAsync/readAsync method
on linux, it's random.. sometime i get the exception after few secondes, sometimes it's 2 hours.

do you have an idea of why i get this problem ?
tried with ping on/off. even tried to write data to the pipe, i don't get exception

The server initialization is: 

```
                var bufferSize = 1024 * 8; 
                var bufferPoolSize = 100 * bufferSize; 
                var options = new WebSocketListenerOptions
                {
                    PingTimeout = TimeSpan.FromSeconds(15),
                    NegotiationTimeout = TimeSpan.FromSeconds(5),
                    PingMode = PingMode.LatencyControl,
                    ParallelNegotiations = 16,
                    NegotiationQueueCapacity = 256,
                    BufferManager = BufferManager.CreateBufferManager(bufferPoolSize, bufferSize)
                };
                options.Transports.ConfigureTcp(tcp =>
                {
                    tcp.BacklogSize = 100;
                    tcp.ReceiveBufferSize = bufferSize;
                    tcp.SendBufferSize = bufferSize;
                    tcp.SendTimeout = TimeSpan.FromSeconds(8);
                    tcp.ReceiveTimeout = TimeSpan.FromSeconds(8);
                    tcp.NoDelay = true; //i send mainly small packets, and the app is realtime.
                });
                options.Standards.RegisterRfc6455();
                server = new WebSocketListener(listenEndPoints, options);
```

the client object read/write/close logic is (part of the code)

```

        public void Start()
        {
            Task taskWrite = new Task(() => WriteAsync());
            taskWrite.Start();
            Task taskRead = new Task(() => ReadAsync());
            taskRead.Start();
            timerLatency.Change(30000, 30000);
        }

        public async Task ReadAsync()
        {
            await Task.Yield();

            try
            {
                while (IsConnected)
                {
                    string msgContent = await sock.ReadStringAsync(CancellationToken.None);
                    WebsocketMessage msg = null;
                    if (msgContent != null)
                    {
                        dateDernierMessageRecu = DateTime.Now;
                        msg = JsonConvert.DeserializeObject<WebsocketMessage>(msgContent);
                        OnMessageReceived?.Invoke(this, new WebsocketClientMessageReceivedEventArgs(this, msg));
                    }
                    else
                    {
                        CloseAsync(true);
                    }
                }
            }
            catch (Exception ex)
            {
                if (Helper.Log.IsErrorEnabled)
                    Helper.Log.Error("Exception while reading data from Websocket #" + connum);
                CloseAsync(false);
            }
        }

        public async Task WriteAsync()
        {
            await Task.Yield();

            while (IsConnected)
            {
                try
                {
                    string data = await outgoingQueue.TakeAsync(); //on prend (ou attend) un message.
                    //Récupère toutes les données à envoyer. on va tout envoyer d'un coup.
                    //on récupère soit tous les paquets, soit un maximum de 32ko de paquet
                    int size = Encoding.UTF8.GetByteCount(data);
                    List<string> toSend = new List<string>();
                    toSend.Add(data);
                    //si on a encore des données, on les ajoute, jusqu'au max de 32ko ou s'il y en a plus.
                    while (size < 32768 && outgoingQueue.Count > 0)
                    {
                        data = await outgoingQueue.TakeAsync();
                        toSend.Add(data);
                        size += Encoding.UTF8.GetByteCount(data);
                    }
                    if (toSend.Count > 0) //puis on envoie.
                    {
                        string realdData = JsonConvert.SerializeObject(toSend, Formatting.None);
                        await sock.WriteStringAsync(realdData, CancellationToken.None);
                    }
                }
                catch (Exception ex)
                {
                    if (Helper.Log.IsErrorEnabled)
                        Helper.Log.Error("Exception while writing data to Websocket #" + connum);
                    CloseAsync(false);
                }
            }
        }

        public async Task CloseAsync(bool normalClear)
        {
            await Task.Yield();

            if (isConnected)
            {
                isConnected = false;
                timerLatency.Change(Timeout.Infinite, Timeout.Infinite);

                int avgLatency = 0;
                if (latencyPackets > 0)
                    avgLatency = (int)(totalLatency / latencyPackets);


                if (normalClear)
                {
                    if (Helper.Log.IsInfoEnabled)
                        Helper.Log.Info("Closing Websocket #" + Connum + " " + Ip + ". Latency (min/max/avg) : " + minLatency + "/" + maxLatency + "/" + avgLatency);
                }
                else
                {
                    if (Helper.Log.IsWarnEnabled)
                        Helper.Log.Warn("Closing Websocket (abnormal close) #" + Connum + " " + Ip + ". Latency (min/max/avg) : " + minLatency + "/" + maxLatency + "/" + avgLatency);
                }

                if (sock != null)
                {
                    try
                    {
                        if (sock.IsConnected)
                            await sock.CloseAsync();
                    }
                    catch { }
                    sock.Dispose();
                }

                dateDeconnexion = DateTime.Now;
                onCloseWebsocketClient?.Invoke(this, null);
            }
        }

```

**Comment by deniszykov** [comment]
Hi @warenbe! Do you have reverse proxy/balancer between client and server? Could you run `netstat -a` and check state of 'hang' connection. Thanks!

**Comment by warenbe** [comment]
Hi @deniszykov 

i got nothing between the server and client
in fact, the ping latency is also always the same (for example 6ms), i would at least expect something like -1 or "error" value...

for my test, i did this: 
launched the server in centos 7 
connected a firefox running on a tablet connected with wifi

here is the netstat showing that the tablet is connected

```
root@IPBX-test: ~ # netstat -an | grep 20000
tcp        0      0 0.0.0.0:20000           0.0.0.0:*               LISTEN
tcp        0     10 192.168.11.150:20000    192.168.11.238:35146    ESTABLISHED

```

And in the log of my server, i see also the connection

`2018-10-11 11:15:00,162 109262 [Thread Pool Worker] INFO  WSServer.WebsocketServer - Nouvelle connexion Websocket #2 : 192.168.11.238:35146`



**now, i disconnect the wifi. so the link is broken.**

in my server log, i don't see the disconnection
even after 1 minute
and in netstat: 

```
root@IPBX-test: ~ # netstat -an | grep 20000
tcp        0      0 0.0.0.0:20000           0.0.0.0:*               LISTEN
tcp        0     50 192.168.11.150:20000    192.168.11.238:35146    ESTABLISHED

```

as you can see, it's like Linux/Mono is not closing the TCP connection... so i don't get any exception when reading/writing to the pipe.
also, the latency calculated by your websocket server is always returning the same value (for example 6 ms )

please note that this occur about 30/40% of the time
sometime it work correctly, and next disconnection is not handled properly

**Comment by warenbe** [comment]
i also tried to lower the keep alive settings in linux, without any succes... :(

http://tldp.org/HOWTO/TCP-Keepalive-HOWTO/usingkeepalive.html

**Comment by deniszykov** [comment]
Actually I can't control how OS detect TCP timeout, but this same function in WS should be performed by the pinging frames. And I found a bug in my ping code. You could try this 0207a7072212cd3a6d2dc6ca1279f586d390a272 commit with fix.

**Comment by warenbe** [comment]
i'll try it as soon as i have 5 minutes.

**Comment by warenbe** [comment]
i'll try it as soon as i have 5 minutes.

**Comment by warenbe** [comment]
Hi
sorry for long delay between tests

so, my problem is still the same: when the ping timeout is reached, the link is not closed.

in your code, this lines should close the connection : 

```
if (this.lastPong.Elapsed > this.pingTimeout)
{
    await this.connection.CloseAsync(WebSocketCloseReasons.GoingAway).ConfigureAwait(false);
    return;
}
```

but it's not the case. the connection remain UP on linux.

i added this in my own code: (some line are checking the time passed since we received a message from the client. the client MUST send a message less than every 60 secondes in my case)

```
TimeSpan delaiDernierMessage = new TimeSpan(DateTime.Now.Ticks - dateDernierMessageRecu.Ticks);
if (delaiDernierMessage.TotalSeconds >= 60)
{
    CloseAsync(false);
}
```

and it works

so there is definitely something bad in the ping timeout result...

**Comment by deniszykov** [comment]
Thanks for feedback. In a84360f6c4f94a1519a4fd42e6ac41bb30ebf048 I threw out the current ping concept based on ping/pong time measurement. And make basic assumption - while incoming packets arrive, the connection can be considered alive. Now I record the time when the incoming packet is received. The ping frame by standard requires an pong response. So even the most calm client will be forced to send respond frames (which time I record and track). So every X seconds I send a ping frame I check how much time has passed since the last packet was received. If elapsed time is greater than PingTimeout I close connection. Also I added log message about timed out connection.

**Comment by warenbe** [comment]
so i made some more tests and added debug log of your code

HEre is what i see when my client get a broken link (i just switch off the wifi of a tablet)

```
2018-10-22 14:43:20,491 70423 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (5D5CC175) [FRAME->] Ping, len: 8, key: 0, flags: FIN
2018-10-22 14:43:28,742 78674 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - WebSocket connection (5D5CC175) has been closed due ping timeout. Time elapsed: 00:00:16.4306358, timeout: 00:00:15.
2018-10-22 14:43:28,743 78675 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (5D5CC175) [FRAME->] ConnectionClose, len: 2, key: 0, flags: FIN
2018-10-22 14:43:32,275 82206 [Thread Pool Worker] INFO  WSServer.WebsocketServer - Websocket #2: latency 0 ms
```


As you can see: your code see the ping timeout
your code close the connection

BUT: in my code, i'm still stucked on this line: 

`string msgContent = await sock.ReadStringAsync();`

correct me but as soon as you're closing the connection for pingtimeout, i should get an exception here right ?

seconf funny thing: if i disconnect myself (after not receiving packet for more than 60 seconds for example), everythink work fine, i get the exception when trying to read.

if i was not clear, tells me.


EDIT: in fact, your code send the connection close frame packet.... and do nothing after ! if we have ping timeout, you should just close the underlying socket , no ?

**Comment by deniszykov** [comment]
> EDIT: in fact, your code send the connection close frame packet.... and do nothing after ! if we have ping timeout, you should just close the underlying socket , no ?

U r right. I'm making graceful disconnection in "dead socket" scenario. It should be `connection.Dispose()` not `connection.CloseAsync()`.

**Comment by deniszykov** [comment]
updated in 527122fac1a141ef09c966d77f32eeea0e45a383

**Comment by warenbe** [comment]
Hi Denis,

I'm sorry, it does not solve the problem
Let me show you the logs: 

```
2018-10-23 09:31:47,864 16522 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - New socket accepted. Remote address: '192.168.11.238:45028', Local address: 192.168.11.150:20000.
2018-10-23 09:31:47,877 16535 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - New client from 'TcpConnection, local: 192.168.11.150:20000, remote: 192.168.11.238:45028, connected: True, available (bytes)529, no delay: False, receive buffer: 8192, receive timeout: 00:00:15.0010000, send buffer: 8192, send timeout: 00:00:15.0010000' is connected.
2018-10-23 09:31:48,169 16826 [Thread Pool Worker] INFO  WSServer.WebsocketServer - Nouvelle connexion Websocket #1 : 192.168.11.238:45028
2018-10-23 09:31:48,278 16936 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (F76ED270) [FRAME<-] Text, len: 117, key: 1B47F5B6, flags: FIN,MASK
2018-10-23 09:31:48,558 17216 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (F76ED270) [FRAME->] Text, len: 203, key: 0, flags:
2018-10-23 09:31:48,579 17237 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (F76ED270) [FRAME->] Continuation, len: 0, key: 0, flags: FIN
2018-10-23 09:31:56,137 24794 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (F76ED270) [FRAME->] Ping, len: 8, key: 0, flags: FIN
2018-10-23 09:31:56,189 24846 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (F76ED270) [FRAME<-] Pong, len: 8, key: 93863C24, flags: FIN,MASK

******* Link was broken here !

2018-10-23 09:32:04,369 33027 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (F76ED270) [FRAME->] Ping, len: 8, key: 0, flags: FIN
2018-10-23 09:32:12,629 41287 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (F76ED270) [FRAME->] ConnectionClose, len: 2, key: 0, flags: FIN
2018-10-23 09:32:18,196 46854 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - Websocket #1: latency 31 ms
2018-10-23 09:32:48,188 76846 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - Websocket #1: latency 31 ms
[....]
```

As you can see, even after ping timeout, you still send the connectionclose packet and wait indefinitely. 
the underlying socket is not disposed. 

another thing: as you can see, i check the latency every 30 seconds. when the link is broken, latency should at least be with an error value like -1 or something like that. at least it would allow me to close manually the connection.
it's not a problem if the server close the socket when ping timeout, which is not the case here.

Edit: the problem seems to be that the dispose method try to close the connection...
```
public void Dispose()
        {
            this.CloseAsync(WebSocketCloseReasons.NormalClose).Wait();

            if (Interlocked.Exchange(ref this.closeState, CLOSE_STATE_DISPOSED) == CLOSE_STATE_DISPOSED)
                return;

            this.options.BufferManager.ReturnBuffer(this.SendBuffer.Array);
            this.options.BufferManager.ReturnBuffer(this.closeBuffer.Array);

            SafeEnd.Dispose(this.writeSemaphore, this.log);
            SafeEnd.Dispose(this.networkConnection, this.log);
        }
```

this is redundant with the closeasync method. why is it here ?

**Comment by deniszykov** [comment]
Seems socket.WriteAsync hung and doesn't return on dead socket. So CloseAsync is never return from writing close frame and socket doesn't get disposed. This is differs from Windows implementation, which doesn't hung on writing to dead socket.

I have removed `this.CloseAsync(WebSocketCloseReasons.NormalClose).Wait();` from Dispose() method, now it is not blocked by writing to a dead socket.

**Comment by warenbe** [comment]
YEs exactly, it's not blocked under windows...

is it normal that the close is also in the ping timeout code before the dispose ?

```
            public override async Task PingAsync()
            {
                if (this.connection.IsClosed)
                    return;

                var elapsedTime = TimestampToTimeSpan(Stopwatch.GetTimestamp()) - this.lastActivityTime;
                if (elapsedTime > this.pingTimeout)
                {

                    await this.connection.CloseAsync(WebSocketCloseReasons.GoingAway).ConfigureAwait(false);
                    SafeEnd.Dispose(this.connection);

                    if (this.connection.log.IsDebugEnabled)
                        this.connection.log.Debug($"WebSocket connection ({this.connection.GetHashCode():X}) has been closed due ping timeout. Time elapsed: {elapsedTime}, timeout: {this.pingTimeout}.");

                    return;
                }

                ((ulong)Stopwatch.GetTimestamp()).ToBytes(this.pingBuffer.Array, this.pingBuffer.Offset);
                var messageType = (WebSocketMessageType)WebSocketFrameOption.Ping;

                var pingFrame = this.connection.PrepareFrame(this.pingBuffer, 8, true, false, messageType, WebSocketExtensionFlags.None);
                await this.connection.SendFrameAsync(pingFrame, TimeSpan.Zero, SendOptions.NoErrors, CancellationToken.None).ConfigureAwait(false);
            }
```

**Comment by deniszykov** [comment]
I also have to mention that sockets in Mono on Linux are not stable. They tend to mix data between instances under heavy load. If you plan to build high load application you should switch to Unix Sockets and reverse proxy. This is one of the reasons why I started this fork. 

> is it normal that the close is also in the ping timeout code before the dispose ?

Already removed.

**Comment by warenbe** [comment]
Thank you for fast update
removing the close in the dispose seems to solve a lot of things: the disconnection is now immediatly detected, no need to wait for ping timeout ;)
```

2018-10-23 10:17:46,375 113929 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (B20470C3) [FRAME->] Continuation, len: 0, key: 0, flags: FIN
2018-10-23 10:17:49,466 117020 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (B20470C3) [FRAME->] Ping, len: 8, key: 0, flags: FIN
2018-10-23 10:17:49,550 117103 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (B20470C3) [FRAME<-] Pong, len: 8, key: D5786D39, flags: FIN,MASK
2018-10-23 10:17:53,865 121419 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (B20470C3) An error occurred while async awaiting header.
2018-10-23 10:17:53,866 121420 [Thread Pool Worker] ERROR WSServer.WebsocketServer - Connection reset by peer
  at vtortola.WebSockets.Rfc6455.WebSocketConnectionRfc6455.AwaitHeaderAsync (System.Threading.CancellationToken cancellation) [0x000ec] in <31d3682387214852885a0060183df2a9>:0
2018-10-23 10:17:53,867 121420 [Thread Pool Worker] DEBUG WSServer.WebsocketServer - (B20470C3) [FRAME->] ConnectionClose, len: 2, key: 0, flags: FIN
2018-10-23 10:17:53,869 121422 [Thread Pool Worker] ERROR WSServer.WebsocketServer - Exception while reading data from Websocket #3
2018-10-23 10:17:53,869 121423 [Thread Pool Worker] WARN  WSServer.WebsocketServer - Closing Websocket (abnormal close) #3 192.168.11.238:45087. Latency (min/max/avg) : 49/49/49

```

**Comment by warenbe** [comment]
> 
> 
> I also have to mention that sockets in Mono on Linux are not stable. They tend to mix data between instances under heavy load. If you plan to build high load application you should switch to Unix Sockets and reverse proxy. This is one of the reasons why I started this fork.
> 
> > is it normal that the close is also in the ping timeout code before the dispose ?
> 
> Already removed.

Well
define heavy load ?

i'm using the server with something like 50/100 sockets opened at a time

**Comment by deniszykov** [comment]
> define heavy load ?

When you open/close/write/read hundreds Sockets per second. 

If you starting to get strange disconnects under load, this should ring a bell.

> removing the close in the dispose seems to solve a lot of things: the disconnection is now immediatly 

Excellent! I will update package soon.

**Comment by warenbe** [comment]
ok. might be the case... 
mono has some strange behavior, not only with sockets btw.

**Comment by deniszykov** [comment]
updated package to [4.2.5](https://www.nuget.org/packages/deniszykov.WebSocketListener/4.2.5)

**Comment by warenbe** [comment]
Hey

i think you can close this issue. 
i'm not having this problem since you updated in october.
